File: MiniProjetAng/src/main/java/tn/esprit/springproject/entities/Foyer.java
package tn.esprit.springproject.entities;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import lombok.*;
import lombok.experimental.FieldDefaults;

import java.io.Serializable;
import java.util.Set;

@Entity
@Getter
@Setter
@ToString(includeFieldNames = false)
@EqualsAndHashCode
@NoArgsConstructor
@AllArgsConstructor
@Builder
@FieldDefaults(level = AccessLevel.PRIVATE)
public class Foyer implements Serializable {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @ToString.Exclude
    @Setter(AccessLevel.NONE)
    Long idFoyer;

    String nomFoyer;

    Long capaciteFoyer;

    @OneToOne
    private Universite universite;


    @OneToMany(mappedBy = "foyer")
    @JsonIgnore
    private Set<Bloc> blocs;
}

File: miniProjet/src/app/model/Foyer.ts
import {Universite} from "./Universite";
import {Bloc} from "./Bloc";

export class Foyer{
  idFoyer: number;
  nomFoyer:String ;
  capaciteFoyer:number;
  universite:Universite;
  bloc:Bloc;



  static calculerMoyenneCapacite(foyers: Foyer[]): number {
    if (foyers.length === 0) {
      return 0;
    }

    const sommeCapacites = foyers.reduce((total, foyer) => total + foyer.capaciteFoyer, 0);
    return sommeCapacites / foyers.length;
  }

  static trouverFoyerCapaciteMaximale(foyers: Foyer[]): Foyer | null {
    if (foyers.length === 0) {
      return null;
    }

    return foyers.reduce((maxFoyer, foyer) => (foyer.capaciteFoyer > maxFoyer.capaciteFoyer ? foyer : maxFoyer), foyers[0]);
  }

}
